src/reporters.ts

```diff
@@ -12,22 +13,41 @@ import { pipe } from "fp-ts/lib/function";
  *   ".a is not a number"
  *   ".c.b is not a non empty string"
  */
+
 const getContextPath = (context: Context): string => {
   if (context.length === 0) {
-    return "] (decoder info n/a)";
+    return " (decoder info n/a)";
   }
-  const keysPath = context.map(({ key }) => key).join(".");
+
+  /*
+    in order to get a more readable ouptut we want to avoid those keys that represent the index of a t.union or a t.intersection, 
+    to do this we want to render the key in the output just in 2 cases: 
+      1. the key can be parsed as a valid integer and the element right before was an ArrayType
+      2. the key is not parsable as a valid integer but it is also a non empty string
+  */
+  const keysPath = context
+    .filter(
+      // eslint-disable-next-line
+      (c: any, i: number) =>
+        (IntegerFromString.is(+c.key) &&
+          context[i - 1] &&
+          // eslint-disable-next-line
+          (context[i - 1].type as any)._tag === "ArrayType") ||
+        (!IntegerFromString.is(+c.key) && c.key !== "")
+    )
+    .map(({ key }) => (!IntegerFromString.is(+key) ? `.${key}` : `[${key}]`));
```

Comment:
What do you think if we rewrite this with this:
```suggestion
  const keysPath = context.reduce(
    (prev: ReadonlyArray<string>, c: ContextEntry, i: number) =>
      c.key !== "" &&
      ((Number.isInteger(+c.key) &&
        i > 0 &&
        Array.isArray(context[i - 1].actual)) ||
        !Number.isInteger(+c.key))
        ? Number.isInteger(+c.key)
          ? [...prev, `[${c.key}]`]
          : [...prev, `.${c.key}`]
        : prev,
    []
  );
```
Enhancement are:
- Use standard Javascript library to verify numbers
- Remove any types
- Check the actual value is an Array instead of manipulate hidden properties of `fp-ts` (`_tag`) that could be change without a compilation error.
- Check always before converting `key` as number that is `!==""` (empty string numeric conversion is `0`)
- collapse `filter` and `map` operations in only one `reduce`

Reply:
> Check the actual value is an Array instead of manipulate hidden properties of fp-ts (_tag) that could be change without a compilation error.

We don't want to check that the _actual_ value is an Array, because the validation is based on what the decoder is expecting to find.

e.g using the following codec

```typescript
t.union([t.string, t.boolean]).decode([1, "2", 3])
```

the returned errors contain the indexes, that we do not expect in this situation

```typescript
["value [1,\"2\",3] at root[0] is not a valid [string]", "value [1,\"2\",3] at root[1] is not a valid [boolean]"]
```

> collapse filter and map operations in only one reduce

I like it, just for making it more readable I would suggest to extract the following code into a function like `isArrayIndex`. 
```typescript 
 c.key !== ""
 && ((Number.isInteger(+c.key) 
 && i > 0 
 &&  (context[i - 1].type as any)._tag === "ArrayType")
```

Reply:
You are right, but this remain a weakness depending on the hidden implementation of `fp-ts`.
Nevertheless maybe only `ArrayType` is not enough, what about other array types (`ReadonlyArrayType`, `AnyArrayType`)?

Reply:
> You are right, but this remain a weakness depending on the hidden implementation of fp-ts.
Nevertheless maybe only ArrayType is not enough, what about other array types (ReadonlyArrayType, AnyArrayType)

Sure, you're right! We should take care of them too, thanks!

Reply:
Nice suggestions I'm applying those changes. Also you are both right about `ReadOnlyArrayType` but
talking about `AnyArrayType` i believe that no check is needed... any type of element inside an `AnyArrayType` would be good for `decode` method (even `null` or `undefined`). The only case where the decode would fail it would be at root if what is passed is not an `Array` but we do not need to specify any index in that case.

Reply:
If you want to avoid the cast to any you can check against `context[i - 1].type.name` ie.

```ts
const isArrayIndex = (c: ContextEntry, i: number, context: Context): boolean =>
  Number.isInteger(+c.key) &&
  i > 0 &&
  includes("Array")(context[i - 1].type.name);

  /*
   * To get a more readable ouptut we avoid to print keys
   * that represent the indexes of a union or a intersection.
   */
  const keysPath = context.reduce(
    (prev: ReadonlyArray<string>, c: ContextEntry, i: number) =>
      isArrayIndex(c, i, context) || !Number.isInteger(+c.key)
        ? Number.isInteger(+c.key)
          ? // key is an integer and the previous element is an array, we keep it
            [...prev, `[${c.key}]`]
          : // key is not an integer, we keep it (dotted notation)
            [...prev, `.${c.key}`]
        : // key is an integer but the previous element is not an array, we skip it
          prev,
    []
  );

```

Moreover, I'd take the lines that build the output outside `isArrayIndex` which should return a boolean.

Reply:
That's a good point, I am going to use `includes` on `type.name` to avoid the cast to `any`,  anyway I think that it's better to leave all the logic inside the `isArrayIndex` function to get it more readable, but I agree that this function should have a different name, something like `getKeysPath`

Reply:
That would be still a bad naming since it doesn't `get` something but instead it writes something (into the output). 
I'm still inclined to put the logic inside the reduce body where it belongs.